add quantized_hardtanh

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34097 add quantized_hardtanh**

Summary:

Adds quantized hardtanh.  Calls the clamp kernel behind the
scenes.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208860](https://our.internmc.facebook.com/intern/diff/D20208860)

vkuzo committed:
add quantized_hardtanh

Summary:

Adds quantized hardtanh.  Calls the clamp kernel behind the
scenes.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

z-a-f commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -41,6 +41,42 @@ Tensor quantized_clamp(
+  Tensor qy;
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "hardtanh", [&]() {
+    qy = quantized_clamp_impl(qx, min, max);
+  });
```
I don't think we need to dispatch here:

```python
qy = quantized_clamp_impl(qx, min, max);
```

should do the trick.

vkuzo replied:
> I don't think we need to dispatch here:

sounds good.  Just for my own understanding, the reason we don't need the dispatch is because there is no need to branch things by `scalar_t`?  Originally I added the dispatch as a copy-pasta from clamp above.

z-a-f replied:
Yea, I think the dispatch was a contamination from earlier versions of the code. But it is still worth testing if we really need it

z-a-f commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -41,6 +41,42 @@ Tensor quantized_clamp(
+    Scalar max) {
+  Tensor qy;
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "hardtanh", [&]() {
+    qy = quantized_clamp_impl(qx, min, max);
```
I think `quantized_clamp_impl` is in the anonymous namespace. Instead, you should use `at::clamp`

vkuzo replied:
[n00b] `qclamp` stub looks like its vectorized using `Vec256`, haven't benchmarked but I'm guessing it might be faster than `at::clamp`?

z-a-f replied:
[native_functions.yaml:L700](https://github.com/pytorch/pytorch/blob/eeda1146a5035bbc019c4dc01ea15966bdd4ab64/aten/src/ATen/native/native_functions.yaml#L700) says that `at::clamp` uses `quantized_clamp`, which in turn [uses `quantized_clamp_impl`](https://github.com/pytorch/pytorch/blob/eeda1146a5035bbc019c4dc01ea15966bdd4ab64/aten/src/ATen/native/quantized/cpu/qclamp.cpp#L39). So I think it should be the same, but without the use of anonymous namespace functions.

z-a-f replied:
Oh, my bad -- in this context you can use either `at::clamp` or `quantized_clamp_impl` -- it will do the same thing. I was confused. My bad!

vkuzo replied:
ah, ic, didn't realize at::clamp is before the dispatch, thanks!

for my own understanding, what's the downside of calling things which live in anonymous in pytorch?

z-a-f replied:
Sorry there is no downside -- I was too sleepy, and thought that you were trying to call an anonymous namespace from another file. This would cause an error. I just misread your code.

z-a-f commented on test/test_quantized.py:
```diff
@@ -252,6 +252,44 @@ def test_qclamp(self, X, min_val, max_val):
+    @given(X=hu.tensor(shapes=hu.array_shapes(1, 8, 1, 8),
+                       elements=hu.floats(-1e6, 1e6, allow_nan=False),
+                       qparams=hu.qparams()),
+           min_val=hu.floats(-1e6, 1e6, allow_nan=False),
```
`allow_infinity=False` See [documentation](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.floats) for more details

z-a-f reviewed (changes requested):
Don't forget to `git submodule --init --recursive` before committing.

vkuzo committed:
Update on "add quantized_hardtanh"


Summary:

Adds quantized hardtanh.  Calls the clamp kernel behind the
scenes.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 reviewed (approved):
LGTM,deferring to zafar for final approval

z-a-f reviewed (approved):
LGTM -- if time permits, can you also add a microbenchmark?

vkuzo committed:
Update on "add quantized_hardtanh"


Summary:

Adds quantized hardtanh.  Calls the clamp kernel behind the
scenes.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208860](https://our.internmc.facebook.com/intern/diff/D20208860)

[ghstack-poisoned]

vkuzo commented:
> LGTM -- if time permits, can you also add a microbenchmark?

done, https://our.internmc.facebook.com/intern/paste/P127140551/ has example results

vkuzo committed:
Update on "add quantized_hardtanh"


Summary:

Adds quantized hardtanh.  Calls the clamp kernel behind the
scenes.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208860](https://our.internmc.facebook.com/intern/diff/D20208860)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized_hardtanh"


Summary:

Adds quantized hardtanh.  Calls the clamp kernel behind the
scenes.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208860](https://our.internmc.facebook.com/intern/diff/D20208860)

[ghstack-poisoned]
